cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/store/profile/ProfileStore.java

```diff
@@ -82,94 +88,152 @@ public static void setupDatasets(DatasetFramework framework) throws IOException,
    *
    * @param profileId the id of the profile to look up
    * @return the profile information about the given profile
-   * @throws IOException if there was an IO error looking up the profile
    * @throws NotFoundException if the profile is not found
    */
-  public Profile getProfile(ProfileId profileId) throws IOException, NotFoundException {
-    return Transactionals.execute(transactional, context -> {
-      Profile profile = getMDS(context).get(getRowKey(profileId), Profile.class);
-      if (profile == null) {
-        throw new NotFoundException(profileId);
-      }
-      return profile;
-    }, IOException.class, NotFoundException.class);
+  public Profile getProfile(ProfileId profileId) throws NotFoundException {
+    return getProfileInfo(profileId).getProfile();
+  }
+
+  /**
+   * Get the profile meta about the given profile, including status and profile property
+   *
+   * @param profileId the id of the profile to look up
+   * @return the profile meta about the given profile
+   * @throws NotFoundException if the profile is not found
+   */
+  public ProfileMeta getProfileMeta(ProfileId profileId) throws NotFoundException {
+    return getProfileInfo(profileId);
   }
 
   /**
    * Get the profile information in the given namespace
    *
    * @param namespaceId the id of the profile to look up
    * @return the list of profiles which is in this namespace
-   * @throws IOException if there was an IO error looking up the profile
    */
-  public List<Profile> getProfiles(NamespaceId namespaceId) throws IOException {
+  public List<Profile> getProfiles(NamespaceId namespaceId) {
     return Transactionals.execute(transactional, context -> {
-      List<Profile> profiles = getMDS(context).list(getRowKey(namespaceId), Profile.class);
-      return Collections.unmodifiableList(profiles);
-    }, IOException.class);
+      List<ProfileMeta> profiles = getMDS(context).list(getRowKey(namespaceId), ProfileMeta.class);
+      return Collections.unmodifiableList(profiles.stream().map(ProfileMeta::getProfile).collect(Collectors.toList()));
+    });
   }
 
   /**
-   * Save the profile to the profile store.
+   * Save the profile to the profile store. By default the profile status will be enabled.
    *
    * @param profileId the id of the profile to save
    * @param profile the information of the profile
-   * @throws IOException if there was an IO error saving the profile
    */
-  public void saveProfile(ProfileId profileId, Profile profile) throws IOException {
+  public void saveProfile(ProfileId profileId, Profile profile) {
     Transactionals.execute(transactional, context -> {
-      getMDS(context).write(getRowKey(profileId), profile);
-    }, IOException.class);
+      getMDS(context).write(getRowKey(profileId), new ProfileMeta(profile, ProfileStatus.ENABLED));
```

Comment:
If somebody modifies an existing profile, this will incorrectly set the status to enabled. You will need to read the existing profile before writing.

Reply:
fixed